sympy/ntheory/ecm.py

```diff
@@ -0,0 +1,250 @@
+#----------------------------------------------------------------------------#
+#                                                                            #
+#                   Lenstra's Elliptic Curve Factorization                   #
+#                                                                            #
+#----------------------------------------------------------------------------#
+
+
+class Point:
+    """For denoting the point on the curve.
+    For ECM we don't need the y-coordinate. All
+    arithmetic operations use only the x and z coordinates.
+    """
+
+    def __init__(self, x, z):
+        self.x = x
+        self.z = z
+
+    @staticmethod
+    def add(P, Q, diff, n):
+        """
+        Add two points P and Q where diff = P - Q.
+
+        Parameters:
+        ===========
+
+        P : point on the curve in Montgomery form
+        Q : point on the curve in Montgomery form
+        diff = P - Q
+        n = modulus
+
+        References
+        ==========
+
+        .. [1]  http://www.hyperelliptic.org/tanja/SHARCS/talks06/Gaj.pdf
+            Algorithm - 3
+        """
+
+        u = (P.x - P.z)*(Q.x + Q.z)
+        v = (P.x + P.z)*(Q.x - Q.z)
+        add, subt = u + v, u - v
+        x_cor = diff.z*add*add % n
+        z_cor = diff.x*subt*subt % n
+        return Point(x_cor, z_cor)
+
+    @staticmethod
+    def double(P, n, a24):
+        """
+        Doubles a point in an elliptic curve in Montgomery form.
+
+        Parameter:
+        ==========
+
+        P : Point on the curve
+        a24 : Parameter for doubling
+        n : modulus
+
+        References
+        ==========
+
+        .. [1]  http://www.hyperelliptic.org/tanja/SHARCS/talks06/Gaj.pdf
+            Algorithm - 3
+        """
+
+        u, v = P.x + P.z, P.x - P.z
+        u, v = u*u, v*v
+        diff = u - v
+        x_cor = u*v % n
+        z_cor = diff*(v + a24*diff) % n
+        return Point(x_cor, z_cor)
+
+    def mont_ladder(self, k, n, a24):
+        """
+        Scalar multiblication of a point in
+        Ellipic Curve in Montgomery form.
+
+        Parameters:
+        ==========
+
+        k : The positive integer multiplier
+        P : Point on the curve in Montgomery form
+        n : modulus
+        a24 : Property of the curve
+
+        References
+        ==========
+
+        .. [1]  http://www.hyperelliptic.org/tanja/SHARCS/talks06/Gaj.pdf
+            Algorithm - 3
+        """
+
+        Q = self
+        R = self.double(self, n, a24)
```

Comment:
I would have defined `add` and `double` as instance methods. Then this line could be `R = Q.double(n, a24)`. (It is not necessary to change the name of the first parameter in `double` to `self`, it can be `P`.)

Reply:
Done. Please take a look.